pkg/diff/checkpoint.go

```
@@ -0,0 +1,373 @@
+	query := fmt.Sprintf("SELECT `chunk_num`, `check_success_num`, `check_failed_num`, `check_ignore_num`, `state` FROM `%s`.`%s` WHERE `schema` = ? AND `table` = ? LIMIT 1",
+		checkpointSchemaName, summaryTableName)
+	rows, err := db.QueryContext(ctx, query, schema, table)
+	if err != nil {
+		return 0, 0, 0, 0, "", errors.Trace(err)
+	}
+	defer rows.Close()
+
+	var totalNum, successNum, failedNum, ignoreNum sql.NullInt64
+	var stateStr sql.NullString
+	for rows.Next() {
+		err1 := rows.Scan(&totalNum, &successNum, &failedNum, &ignoreNum, &stateStr)
+		if err1 != nil {
+			return 0, 0, 0, 0, "", errors.Trace(err1)
+		}
+
+		if !totalNum.Valid || !successNum.Valid || !failedNum.Valid || !ignoreNum.Valid || !stateStr.Valid {
+			return 0, 0, 0, 0, "", errors.New("get summary failed, value is valid")
+		}
+
+		return totalNum.Int64, successNum.Int64, failedNum.Int64, ignoreNum.Int64, stateStr.String, nil
+
+	}
+	if rows.Err() != nil {
+		return 0, 0, 0, 0, "", errors.Trace(rows.Err())
+	}
+
+	return 0, 0, 0, 0, "", errors.NotFoundf("schema %s, table %s summary info", schema, table)
+}
+
+// getChunkSummary get the table's summary info from `chunk` table
+func getChunkSummary(ctx context.Context, db *sql.DB, instanceID, schema, table string) (total, successNum, failedNum, ignoreNum int64, err error) {
+	query := fmt.Sprintf("SELECT `state`, COUNT(*) FROM `%s`.`%s` WHERE `instance_id` = ? AND `schema` = ? AND `table` = ? GROUP BY `state` ;", checkpointSchemaName, chunkTableName)
+	rows, err := db.QueryContext(ctx, query, instanceID, schema, table)
+	if err != nil {
+		return 0, 0, 0, 0, errors.Trace(err)
+	}
+	defer rows.Close()
+
+	var chunkState sql.NullString
+	var num sql.NullInt64
+	for rows.Next() {
+		err1 := rows.Scan(&chunkState, &num)
+		if err1 != nil {
+			return 0, 0, 0, 0, errors.Trace(err1)
+		}
+
+		if !chunkState.Valid || !num.Valid {
+			continue
```

do not need to return an error?